### PR TITLE
implement methods and request methods in WoodokuBoardRepresentation

### LINKED ISSUE #5: Finish implementing class WoodokuBoardRepresentation

### LINKED ISSUE #20: Request changes and additions to WoodokuRepresentation
Request implementation to 3 additional class methods (`is_occupied` is here for reference)

```python
    def is_occupied(self, blocks_coord: List[Tuple[int, int]]) -> bool:
        """Check if each block has is occupied. If all of those blocks are
        occupied, return true. Otherwise, false.

        Args:
            blocks_coord (List[Tuple[int, int]]): list of blocks to check

        Returns:
            bool: if all blocks in `blocks_coord` is occupied
        """

    def is_not_occupied(self, blocks_coord: List[Tuple[int, int]]) -> bool:
        """Check if each block is empty. If all of those blocks are empty,
        return true. Otherwise, false.

        Args:
            blocks_coord (List[Tuple[int, int]]): list of blocks to check

        Returns:
            bool: if all blocks in `blocks_coord` is empty
        """
    
    def __validate(block: Tuple[int, int]) -> None:
        """validate if block is within the 9x9 board. raise Error if not.

        Args:
            block (Tuple[int, int]): The block to be validated

        Raises:
            ShapeOutOfBoardError: when some block is not valid 
        """
        pass
    
    
    def __str__(self) -> str:
        raise NotImplementedError()
```

### PR DESCRIPTION
**Pull request by YANLING-H**

### CODE REVIEW

File: Woodoku/src/Entity/WoodokuBoard.py [hunk 1]
```diff
@@ -12,17 +14,36 @@ class WoodokuBoardRepresentation:
...
+    board: ndarray
+    length: int
+
+    def __init__(self, n: int):
+        self.length = n
+        self.board = np.zeros((self.length, self.length))
 
     def __str__(self):
-        pass
+        # reformat = np.reshape(self.board, self.length * self.length)
```
Comment (EdwardHaoranLee) [inline Woodoku/src/Entity/WoodokuBoard.py]:
```suggestion

```

File: Woodoku/src/Entity/WoodokuBoard.py [hunk 2]
```diff
@@ -31,7 +52,10 @@ def is_occupied(self, blocks_coord: List[Tuple[int, int]]) -> bool:
         :param blocks_coord:
         :return:
         """
-        pass
+        for position in blocks_coord:
+            if self.board[position[0]][position[1]] == 0:
+                return False
+        return False
```
Comment (EdwardHaoranLee) [inline Woodoku/src/Entity/WoodokuBoard.py]: Always returns False?

**Review by EdwardHaoranLee — CHANGES_REQUESTED** [review]

File: Woodoku/src/Entity/WoodokuBoard.py [hunk 3]
```diff
@@ -12,17 +14,36 @@ class WoodokuBoardRepresentation:
     The first block on the bottom row is (n - 1, 0).
     The bottom-right block's coordinate is (n - 1, n - 1).
     """
-    def __init__(self):
-        pass
+    board: ndarray
+    length: int
+
+    def __init__(self, n: int):
+        self.length = n
```
Comment (Xiaorui-Huang) [inline Woodoku/src/Entity/WoodokuBoard.py]: why `n` ? shouldn't it always be 9?

File: Woodoku/src/Entity/WoodokuBoard.py [hunk 4]
```diff
@@ -12,17 +14,36 @@ class WoodokuBoardRepresentation:
...
+        self.board = np.zeros((self.length, self.length))
```
Comment (Xiaorui-Huang) [inline Woodoku/src/Entity/WoodokuBoard.py]:
```suggestion
self.board = np.full((self.length, self.length), False)
```
Document what 0 and 1 means for the board in the docstring of the class
Don't think it matters too much, but a 2d array of bool might be preferable with less memory footprint.

File: Woodoku/src/Entity/WoodokuBoard.py [hunk 5]
```diff
@@ -12,17 +14,36 @@ class WoodokuBoardRepresentation:
...
 
     def __str__(self):
-        pass
+        # reformat = np.reshape(self.board, self.length * self.length)
+        return np.array2string(self.board)
```
Comment (Xiaorui-Huang) [inline Woodoku/src/Entity/WoodokuBoard.py]: this is a little too primitive and not quite visualizable with just 1 and 0

File: Woodoku/src/Entity/WoodokuBoard.py [hunk 6]
```diff
@@ -12,17 +14,36 @@ class WoodokuBoardRepresentation:
...
     def add_blocks(self, blocks_coord: List[Tuple[int, int]]) -> None:
-        pass
+        """
+        Mark each position specified in blocks_coord as one to indicate that the position is occupied.
+
+        :param blocks_coord: a list of (x,y) tuples to be added to the board
+        :return:
+        """
+        for position in blocks_coord:
+            self.board[position[0]][position[1]] = 1
```
Comment (Xiaorui-Huang) [inline Woodoku/src/Entity/WoodokuBoard.py]:
```suggestion
        for row, col in blocks_coord:
            self.board[row, col] = 1
```

File: Woodoku/src/Entity/WoodokuBoard.py [hunk 7]
```diff
@@ -12,17 +14,36 @@ class WoodokuBoardRepresentation:
...
     def remove_blocks(self, blocks_coord: List[Tuple[int, int]]) -> None:
-        pass
+        """
+        Mark each position specified in blocks_coord as zero to indicate that the position is not occupied.
+
+        :param blocks_coord: a list of (x,y) tuples to be added to the board
+        :return:
+        """
+        for position in blocks_coord:
+            self.board[position[0]][position[1]] = 0
```
Comment (Xiaorui-Huang) [inline Woodoku/src/Entity/WoodokuBoard.py]:
```suggestion
        for row, col in blocks_coord:
            self.board[row, col] = 0
```

File: Woodoku/src/Entity/WoodokuBoard.py [hunk 8]
```diff
@@ -31,7 +52,10 @@ def is_occupied(self, blocks_coord: List[Tuple[int, int]]) -> bool:
         :param blocks_coord:
         :return:
         """
-        pass
+        for position in blocks_coord:
+            if self.board[position[0]][position[1]] == 0:
```
Comment (Xiaorui-Huang) [inline Woodoku/src/Entity/WoodokuBoard.py]:
```suggestion
        for row, col in blocks_coord:
            if self.board[row, col] == 0:
```

File: Woodoku/src/Entity/WoodokuBoard.py (hunk 3 above)
Comment (EdwardHaoranLee) [reply Woodoku/src/Entity/WoodokuBoard.py]: I agree that we could have this flexibility.

File: Woodoku/src/Entity/WoodokuBoard.py (hunk 4 above)
Comment (EdwardHaoranLee) [reply Woodoku/src/Entity/WoodokuBoard.py]: Agreed.

File: Woodoku/src/Entity/WoodokuBoard.py (hunk 5 above)
Comment (EdwardHaoranLee) [reply Woodoku/src/Entity/WoodokuBoard.py]: This method was reserved to use in `CommandLineUI`. We might need a more detailed design on the print out format. For now we could leave this method blank.

File: Woodoku/src/Entity/WoodokuBoard.py (hunk 3 above)
Comment (EdwardHaoranLee) [reply Woodoku/src/Entity/WoodokuBoard.py]: You know what, it will make the 3x3 box size be a pain-in-the-ass if we have flexible board size. So we might just fix the number.

File: Woodoku/src/Entity/WoodokuBoard.py (hunk 2 above)
Comment (YANLING-H) [reply Woodoku/src/Entity/WoodokuBoard.py]: I will change it

File: Woodoku/src/Entity/WoodokuBoard.py (hunk 1 above)
Comment (YANLING-H) [reply Woodoku/src/Entity/WoodokuBoard.py:25]: Is there specific string format of the board you need?

File: Woodoku/src/Entity/WoodokuBoard.py (hunk 3 above)
Comment (YANLING-H) [reply Woodoku/src/Entity/WoodokuBoard.py]: I see, I will fix the size to be 9

File: Woodoku/src/Entity/WoodokuBoard.py (hunk 4 above)
Comment (YANLING-H) [reply Woodoku/src/Entity/WoodokuBoard.py]: I see, I will change it

File: Woodoku/tests/Entity/test_woodoku_board.py [hunk 9]
```diff
@@ -0,0 +1,20 @@
...
+    def test_add_blocks_add_multiple_blocks(self):
+        pass
+
+    def test_remove_blocks_remove_one_block(self):
+        pass
+
+    def test_remove_blocks_remove_multiple_block(self):
+        pass
+
+
```
Comment (EdwardHaoranLee) [inline Woodoku/tests/Entity/test_woodoku_board.py]: Should test cases be added in a separate PR?

**Review by EdwardHaoranLee — DISMISSED** [review]

File: Woodoku/tests/Entity/test_woodoku_board.py (hunk 9 above)
Comment (Xiaorui-Huang) [reply Woodoku/tests/Entity/test_woodoku_board.py]: Yeah, test should go in another PR

Comment (YANLING-H) [reply Woodoku/tests/Entity/test_woodoku_board.py]: I see, I will remove the test for now

File: Woodoku/src/Entity/WoodokuBoard.py [hunk 10]
```diff
@@ -13,16 +16,37 @@ class _WoodokuBoardRepresentation:
...
+    _board: an 2d array to record the occupancy of each position on the game board. The value is set to True when
+    the position occupied
     """
+    __board: ndarray
 
     def __init__(self):
+        self.__board = np.full((N, N), False)
+
+    def __str__(self) -> str:
         pass
```
Comment (Xiaorui-Huang) [inline Woodoku/src/Entity/WoodokuBoard.py]: Style comment: all __someting__ function except __init__ should the last methods of the class

**Comment by Xiaorui-Huang** [comment]
Remember to link the related issue if you have changed something out of the scope of the current issues

File: Woodoku/src/Entity/WoodokuBoard.py [hunk 11]
```diff
@@ -55,11 +84,11 @@ def __validate(block: Tuple[int, int]) -> None:
         Raises:
             ShapeOutOfBoardError: when some block is not valid 
         """
-        pass
-    
-    
+        if not (0 <= block[0] <= N - 1 and 0 <= block[1] <= N - 1):
+            raise ShapeOutOfBoardError(block[0], block[1])
```
Comment (Xiaorui-Huang) [inline Woodoku/src/Entity/WoodokuBoard.py]:
```suggestion
		x, y = block
        if not (0 <= x <= N - 1 and 0 <= y <= N - 1):
            raise ShapeOutOfBoardError(x, y)
```
As a side note, try not to do index access in python for tuples. Unpack them. Here is another trick. `*block` is the same as `x, y`

to indeed you can also write line 89 as `raise ShapeOutOfBoardError(*block)`

**Comment by Xiaorui-Huang** [comment]
> Remember to link the related issue if you have changed something out of the scope of the current issues

And also update the PR title/description

File: Woodoku/src/Entity/WoodokuBoard.py (hunk 11 above)
Comment (YANLING-H) [reply Woodoku/src/Entity/WoodokuBoard.py]: I see, but for readability, I might keep using x,y when raising the error

**Comment by YANLING-H** [comment]
Some format in the WoodokuBoard class is changed due to auto-formating in pycharm. I will disable this feature in the next pr.

**Review by Xiaorui-Huang — APPROVED** [review]
